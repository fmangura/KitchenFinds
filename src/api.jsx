import React, {useState} from 'react'
import axios from 'axios'
import qs from 'qs';


import Cookies from 'universal-cookie';
const cookies = new Cookies(null, { path: '/' });

const BACKEND_API_URI = 'http://localhost:3001';

class backend {  
    static user_token = cookies.get('user_token') || ""

    // General request to automatically include token 

    static async request(endpoint, data = {}, method = "GET") {
        const url = `${BACKEND_API_URI}${endpoint}`;
        const headers = { authorization: `Bearer ${backend.user_token}`};
        const params = (method === "GET")
            ? data
            : {};

        try {
            return (await axios({ url, method, data, params, headers })).data;
        } catch (err) {
            console.error("Backend API Error:", err.response);
        }
    }

    // Users individual routes

    static async register(data) {
        return await backend.request(`/users/register`, data, 'post').then((res => {message: 'Register Complete!'}))
    }

    static async login(formData) {
        try {
            return await this.request(`/users/login`, formData, 'post')
                .then((res) => {
                    let {token, user} = res
                    cookies.set('user_token', token)
                    cookies.set('user', user)
                    backend.user_token = cookies.get('user_token') || ""
                    return user
                })
        } catch (err) {
            console.log(err.data)
            return err.data
        }
    }

    static async delUser(formData) {
        try {
            return await this.request(`/users/${cookies.get('user')}`, formData, 'delete')
                .then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getUserRecipes() {
        try{
            return await backend.request(`/users/myrecipes`).then((res) => res)
        } catch (err) {
            return err
        }
    }

    // Recipes individual routes

    static async getRecipeIngredients(ref_id) {
        return await backend.request(`/recipes/getRecipe/${ref_id}`)
    }

    static async searchByIngredients(items) {
        try {
            const ingrList = qs.stringify({ ingredients: items }, { arrayFormat: 'comma' })
            return await axios({
                url: `${BACKEND_API_URI}/recipes/search?${ingrList}`,
                method: "GET",
                headers: { authorization: `Bearer ${backend.user_token}`},
            }).then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getInstructions(ref_id) {
        try {
            return await this.request(`/recipes/instructions/${ref_id}`).then((res) => {console.log(res); return res})
        } catch (err) {
            return err
        }
        
    }

    static async addToFave(ref_id) {
        try {
            return await backend.request(`/recipes/${ref_id}/fave`, {},'post').then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async delFave(ref_id) {
        try {
            return await backend.request(`/recipes/${ref_id}`, {}, 'delete').then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getSuggestions(search) {
        try {
            return await axios({
                url:`${BACKEND_API_URI}/recipes/suggestions`, 
                data: {'item': search},
                method:'POST'
            }).then((res) => {console.log(res.data); return res.data})
        } catch (err) {
            return err
        }
    }
}

export default backend;

