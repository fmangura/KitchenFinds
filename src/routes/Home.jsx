import React, {useContext, useState, useEffect} from 'react'
import {TfiClose} from 'react-icons/tfi';
import { Navigate } from 'react-router-dom';
import { AuthContext } from '../context';
import IngredientsTag from '../components/IngredientTag';
import Recipes from '../components/RecipesOverlay';
import Recipe from '../components/Recipe';
import backend from '../api';
import './home.css'
import './homeLess.less'
import Logo from '../assets/KitchenWordLogo.svg'
import Rat from '../assets/rat.svg'

export default function Home () {
    const {currUser, isLoggedIn, logout, faveRecipes} = useContext(AuthContext)
    const [ingredientsList, setIngredientList] = useState({})
    const [search, setSearch] = useState('')
    const [recipes, setRecipes] = useState()
    const [activeRecipe, setActiveRecipe] = useState()
    const [error, setError] = useState('')
    const [hidden, setHidden] = useState(true)
    const [suggest, setSuggest] = useState([])
    const [slide, setSlide] = useState('')
    const [faves, setFaves] = useState(async () => await backend.getUserRecipes())

    if (!isLoggedIn) {
        return <Navigate to={'/login'}/>
    }

    const handleSearch = (e) => {
        const {value} = e.target;
        setSearch(value)
    }

    const addIngredient = (e) => {
        e.preventDefault()
        if (!search) return
        changeSuggestions()
        setIngredientList(ingredientsList => ({...ingredientsList, [search]: null}))
        setSearch('')
    }

    const pushSuggested = (word) => {
        setIngredientList(ingredientsList => ({...ingredientsList, [word]: null}))
    }

    const changeSuggestions = async () => {
        try {
            await backend.getSuggestions(search).then((res) => {setSuggest(suggest => [...res])})
        } catch (err) {

        }
    }

    const rmIngredient = (word) => {
        let newList = {...ingredientsList};
        delete newList[word];
        setIngredientList(newList);
    }

    const recipeSearch = async () => {
        try {
            let ingredients = Object.keys(ingredientsList)
            let res = await backend.searchByIngredients(ingredients)
            if (res.status != 200) throw Error
            setRecipes(recipes => [...res.data])
            setHidden(null)
        } catch (err) {
            setError("Sorry, cannot reach the cupboard please try again later! ...we'll need a step-ladder")
        }
    }

    const actRecipe = (recipe) => {
        setActiveRecipe(recipe)
    }

    const rmActRecipe = () => {
        setActiveRecipe(activeRecipe => {})
    }

    const reset = () => {
        setIngredientList(ingredientsList => {return {} })
        setSuggest(suggest => {return [] } )
        setSearch('')
    }

    const listSlider = () => {
        setSlide('slide')
    }

    return (
        <div className='fullBody'>  
            <div className='left-cards'>
                {recipes ? 
                    <div className={`recipes-list-overlay ${slide}`} style={hidden ? {display:'none'} : {}}>
                        {recipes.length > 0 ? 
                            <div className='recipe-list'>
                                {recipes.map((recipe) => {return <Recipes key={recipe.id} id={recipe.id} image={recipe.img} name={recipe.name} ingredients={recipe.allIngr} searchedFor={ingredientsList} actRecipe={actRecipe} listSlider={listSlider} activeRecipe={activeRecipe}/>})} 
                            </div> 
                            :
                            <div className='card' style={{justifySelf:'center'}}>
                                <h2>Could not find recipes with these ingredients</h2>
                            </div>
                        }
                    </div>
                : null}
                <div className='main' onClick={() => {
                    (hidden ? null : setHidden(prev => !prev));
                    (setSlide(''));
                    }}>
                    <div className='card--inverted card'>
                        <form onSubmit={addIngredient}>
                        <h2>Add Ingredients: </h2>
                        <label htmlFor="search" className='input'>
                            <input type="text" className="input__field" name='ingredient' placeholder=' ' onChange={handleSearch} value={search}/>
                            <span className="input__label">Input ingredients...</span>
                        </label>
                        <div className="button-group">
                            <button className='addButton'>Add</button>
                            </div>
                        </form>
                    </div>
                    <div className='suggestions card'>
                        <h2 style={{marginRight:'20px'}}>Suggestions: </h2>
                        <ul id='suggested-words'>
                            {suggest ? suggest.map((word) => {
                                return <IngredientsTag key={word} ingredient={word} pushSuggested={pushSuggested} suggest={true} status='clickable'/>}) : null}
                        </ul>
                    </div>
                    <div className='card list-of-ingredients card--accent'>
                        <h2>Ingredients: </h2>
                        <ul>
                            {Object.keys(ingredientsList) ? Object.keys(ingredientsList).map((ingr) => {return <IngredientsTag key={ingr} rmIngredient={rmIngredient} ingredient={ingr} status='clickable'/>}) : null } 
                        </ul>
                    </div>
                    <button style={{marginTop:'20px'}} onClick={recipeSearch} className='cookButton'>Cook!</button>
                    <button type="reset" onClick={reset}>Clear</button>
                    {error ? <p>{error}</p> : null}
                </div>
            </div>
            <div className='right-cards'>
                <div className='infocard card'>
                    {activeRecipe ? 
                        <div className='cook-book'>
                            <button onClick={rmActRecipe} id='close'><TfiClose/></button>
                            <Recipe id={activeRecipe.id} name={activeRecipe.name} image={activeRecipe.image} allIngredients={activeRecipe.ingredients} recipeInfo={activeRecipe.recipeInfo} />
                        </div> 
                        :
                        <>
                            <img src={Logo} alt="" />
                            <p style={{margin:'20px'}}>
                                Ever looked in your fridge after forgetting to go grocery shopping? Don't have time to go grocery shopping? Want to try figuring out a meal? Try looking here!
                                <br/>
                                <br/>
                                For finding recipes that utilizes what you have left in the fridge. Just simply add the ingredients one-by-one into the list and cook!
                                <br/> 
                                <br/> 
                                Make sure you're checking your expiration dates! ðŸ˜…
                            </p>
                        </>
                    }
                    </div>
                <img src={Rat} alt="" style={{height:'100px', width:'100px'}}/>
            </div>
        </div>
    );
}