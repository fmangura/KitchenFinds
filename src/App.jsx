import React, {useState} from 'react'
import { BrowserRouter, Routes, Route } from 'react-router-dom';

import Auth from './components/AuthProvider'
import Home from './routes/Home'
import LoginForm from './routes/LoginForm'
import RegisterForm from './routes/RegisterForm'
import Nav from './components/Nav';
import Profile from './routes/Profile'

export default function App () {

    return (
        <>
            <BrowserRouter>
                <Auth>
                    <Nav />
                    <Routes>
                        <Route path='/' element={<Home />} />
                        <Route path='/profile' element={<Profile />} />
                        <Route path='/login' element={<LoginForm />} />
                        <Route path='/register' element={<RegisterForm />} />
                    </Routes>
                </Auth>
            </BrowserRouter>
        </>
    );
}