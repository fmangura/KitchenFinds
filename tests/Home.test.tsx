import React from 'react';
import { AuthContext } from '../src/context';
import {fireEvent, getByText, queryAllByText, render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import Home from '../src/routes/Home'
import backend from '../src/api';
import { vi } from 'vitest';

vi.mock('../src/api')

const renderComponent = () => {
    const value = {currUser:'u1', isLoggedIn:'true'}
    return render(
        <AuthContext.Provider value={value} >
            <Home/>
        </AuthContext.Provider>
    );
}

describe('Home', () => {
    it('renders without rerouting', () => {
        renderComponent();
        const button = screen.getByRole('button', {name:'Cook!'})
        expect(button).toBeInTheDocument()
    });

    it('checks snapshot', () => {
        const {asFragment} = renderComponent();
        expect(asFragment).toMatchSnapshot();
    })

    it('able to input and add', async () => {
        backend.getSuggestions = vi.fn().mockImplementation(()=>['spinach', 'fries'])
        renderComponent();
        const ingrInput = screen.getByRole('textbox',{name:''});
        const add = screen.getByRole('button', {name:'Add'});
        fireEvent.change(ingrInput, {target:{value:'spinach'}});
    })

})