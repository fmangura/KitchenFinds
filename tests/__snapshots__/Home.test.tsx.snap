// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > checks snapshot 1`] = `[Function]`;
